tgintegration/collector.py

```diff
@@ -0,0 +1,174 @@
+
+
+@dataclass
+class Expectation:
+    min_messages: Union[int, NotSet] = NotSet
+    max_messages: Union[int, NotSet] = NotSet
+
+    def is_sufficient(self, messages: List[Message]) -> bool:
+        n = len(messages)
+        if self.min_messages is NotSet:
+            return n >= 1
+        return n >= self.min_messages
+
+    def _is_match(self, messages: List[Message]) -> bool:
+        n = len(messages)
+        return (self.min_messages is NotSet or n >= self.min_messages) and (
+            self.max_messages is NotSet or n <= self.max_messages
+        )
+
+    def verify(self, messages: List[Message], timeouts: TimeoutSettings) -> None:
+        if self._is_match(messages):
+            return
+
+        n = len(messages)
+
+        if n < self.min_messages:
+            raise_or_log(
+                timeouts,
+                "Expected {} messages but only received {} after waiting {} seconds.",
+                self.min_messages,
+                n,
+                timeouts.max_wait,
+            )
+            return
+
+        if n > self.max_messages:
+            raise_or_log(
+                timeouts,
+                "Expected only {} messages but received {}.",
+                self.max_messages,
+                n,
+            )
+            return
+
+
+@asynccontextmanager
+async def collect(
+    controller: "BotController",
+    filters: Filter = None,
+    expectation: Expectation = None,
+    timeouts: TimeoutSettings = None,
+) -> AsyncContextManager[Response]:
+    expectation = expectation or Expectation()
+    timeouts = timeouts or TimeoutSettings()
+
+    recorder = MessageRecorder()
+    handler = MessageHandler(recorder.record_message, filters=filters)
+
+    async with add_handler_transient(controller.client, handler):
+        response = Response(controller, recorder)
+
+        logger.debug("Collector set up. Executing user-defined interaction...")
+        yield response  # Start user-defined interaction
+        logger.debug("interaction complete.")
+
+        num_received = 0
+        last_received_timestamp = (
+            None  # TODO: work with the message's timestamp instead of utcnow()
+        )
+        timeout_end = datetime.utcnow() + timedelta(seconds=timeouts.max_wait)
+
+        try:
+            seconds_remaining = (timeout_end - datetime.utcnow()).total_seconds()
+
+            while True:
+                if seconds_remaining > 0:
+                    # Wait until we receive any message or time out
+                    logger.debug(f"Waiting for message #{num_received + 1}")
+                    await asyncio.wait_for(
+                        recorder.wait_until(
+                            lambda msgs: expectation.is_sufficient(msgs)
+                            or len(msgs) > num_received
+                        ),
+                        timeout=seconds_remaining,
+                    )
+
+                num_received = len(recorder.messages)  # TODO: this is ugly
```

Comment:
Why is it ugly?

Reply:
- [ ] Add more info about a better architecture